feat: use typescript for the core package

domoritz opened the pull request:
Fixes #312 

Update the vite config to use the source files so we don't need to run tsc manually. I migrated the code in core and mostly kept things as is.

jheer commented on packages/core/src/Selection.ts:
```diff
@@ -28,7 +28,7 @@ export function isSelection(x: any): x is Selection {
 }
 
-function create(options: SelectionResolverOptions, include?: Selection | Selection[]): Selection {
+function create(options: SelectionResolverOptions, include?: Selection[]): Selection {
```
Why the change in types here? The code was designed to accept either a single selection or an array. And if we make the change, the code below could be adjusted to avoid the array wrap and `flat` call.

domoritz replied:
Ah, I misinterpreted the constructor and logic here. I changed this to `include ? [include].flat() : undefined` so that the code typechecks correctly.

domoritz commented:
@kwonoh I know you have experience with TypeScript so I'd appreciate your review.

kwonoh commented on packages/core/src/QueryConsolidator.ts:
```diff
@@ -240,29 +268,27 @@ async function processResults(group, cache) {
+function filterResult(data: Table, map: Array<[string, string]>): any[] {
   const lookup = new Map(map);
-  const result = [];
+  const result: any[] = [];
```
I'm encountering the below error when building one of my projects that has `"skipLibCheck": true` using this branch:
```
../../../../repo/mosaic/packages/core/src/QueryConsolidator.ts:288:19 - error TS2763: Cannot iterate value because the 'next' method of its iterator expects type 'null', but for-of will always send 'undefined'.

288   for (const d of data) {
                      ~~~~
```

It seems to originate from this line in the flechette repo:
https://github.com/uwdata/flechette/blob/158477706cb8e0268a081b190989e706f9f6437d/src/table.js#L158

I will open a PR to the flechette repo after I have better understanding of the `Generator` type

cc @jheer

domoritz commented on tsconfig.base.json:
```diff
@@ -1,13 +1,19 @@
 
+    // disable some checks for now until we have stricter types
+    "noImplicitAny": false,
+    "strictNullChecks": false,
```
I agree. I want to do this incrementally. I enabled stictNullCheck for the core package. I thought `strict:true` would enable it but it seems like the ts options are not as smart and only allow overriding specific properties at a time.

I enabled strict type checks for core and disabled it at the top level for now. Once this is merged, it would be awesome if you could help move package by package over and enable `strict: true` over time.

domoritz commented:
@kwonoh I addressed your comments and also made the types for core strict. We can move the rest incrementally but this pull request should set things up.
